luarules/gadgets/unit_custom_weapons_cluster.lua

```diff
@@ -59,281 +58,258 @@ local cos   = math.cos
-    -- That mostly means not checking and not rotating the unit's collider around in world space.
-    local colliders = spGetUnitsInSphere(ex, ey, ez, 80)
-    local udefid, bounce, ux, uy, uz, uw, radius
-    for _, uid in ipairs(colliders) do
-        udefid = spGetUnitDefID(uid)
-        bounce = unitBulk[udefid]
-        if bounce ~= nil then
-            -- Assuming spherical collider in frictionless vacuum
-            _,_,_,ux,uy,uz = spGetUnitPosition(uid, true)
-            radius         = spGetUnitRadius(uid)
-            if uy + radius > 0 then
-                ux, uy, uz = ex-ux, ey-uy, ez-uz
-                distance   = ux*ux + uy*uy + uz*uz -- just going to reuse this var a lot
-                uw         = distance / radius
-                distance   = sqrt(distance)
-
-                -- We allow wiggle room since our colliders are not spheres
-                if uw <= 1.24 * distance then
-                    distance = max(1, distance / radius)
-                    -- Even with a bunch of transcendentals, the perf isn't so bad.
-                    local th_z = atan2(ux, uz)
-                    local ph_y = atan2(uy, sqrt(ux*ux+uz*uz))
-                    local cosy = cos(ph_y)
-                    x = x + bounce / distance * sin(th_z) * cosy
-                    y = y + bounce / distance * sin(ph_y)
-                    z = z + bounce / distance * cos(th_z) * cosy
-                end
-            end
-        end
-    end
-    return { x, y, z }
+--------------------------------------------------------------------------------
+-- Functions -------------------------------------------------------------------
+
+local function GetSurfaceDeflection(data, projectileID, ex, ey, ez)
+	-- Deflection from deep water, shallow water, and solid terrain.
+	local elevation = spGetGroundHeight(ex, ez)
+	local separation
+	local dx, dy, dz
+	if elevation < deepWaterDepth then
+		separation = ey - deepWaterDepth * (1/2.4)
+		dx = 0
+		dy = 1
+		dz = 0
+	else
+		separation = ey - elevation
+		local slope
+		dx, dy, dz, slope = spGetGroundNormal(ex, ez, true)
+		if slope > 0.1 or slope * separation > 10 then
+			separation = separation * cos(slope)
+			local shift = separation * sin(slope) / sqrt(dx*dx + dz*dz)
+			local sx = ex - dx * shift -- Next surface x, z
+			local sz = ez - dz * shift
+			elevation = max(elevation, spGetGroundHeight(sx, sz))
+			separation = ey - elevation
+			dx, dy, dz = spGetGroundNormal(sx, sz, true)
+		end
+		if elevation <= 0 then
+			separation = ey - max(elevation * (1/1.8), deepWaterDepth * (1/2.4))
+			dx = dx * 0.94
+			dz = dz * 0.94
+			dy = dy < 0.94 and dy / sqrt(dx*dx + dz*dz) * (1/0.94) or 0.94
+		end
+	end
+	separation = 1.3 / sqrt(max(1, separation))
+	dx = dx * separation
+	dy = dy * separation
+	dz = dz * separation
+
+	-- Additional deflection from units, from none to solid-terrain-like.
+	local unitsNearby = spGetUnitsInSphere(ex, ey, ez, 270/2) -- gettin yuge (air repair pad size)
```

Comment:
How is airpad relevant?

Reply:
I assume it has the largest collision volume of any unit, and this code deals with "bouncing" clusters off massive units, which actually nerfs cluster weapons when you think about it.

Reply:
Consider
```lua
local function GetLargestUnitRadius()
  -- Not finding the largest unit programmatically because modding will never exist anyway
  return 270/2
end
```